src/Spago/Config.purs

```diff
@@ -174,130 +175,212 @@ type ReadWorkspaceOptions =
 isRootPackage :: WorkspacePackage -> Boolean
 isRootPackage p = Path.localPart p.path == ""
 
--- | Reads all the configurations in the tree and builds up the Map of local
--- | packages to be integrated in the package set
-readWorkspace :: ∀ a. ReadWorkspaceOptions -> Spago (Registry.RegistryEnv (rootPath :: RootPath | a)) Workspace
-readWorkspace { maybeSelectedPackage, pureBuild, migrateConfig } = do
-  { rootPath } <- ask
+spagoYaml = "spago.yaml" :: String
+
+discoverWorkspace :: ∀ a. ReadWorkspaceOptions -> GlobalPath -> Spago (Registry.RegistryEnv a) { workspace :: Workspace, rootPath :: RootPath }
+discoverWorkspace options cwd = do
   logInfo "Reading Spago workspace configuration..."
+  logDebug $ "Discovering nearest workspace " <> spagoYaml <> " starting at " <> Path.quote cwd
+
+  { workspace, rootPath } /\ upTree@{ closestPackage } <-
+    State.runStateT (walkDirectoriesUpFrom cwd)
+      { loadedPackages: Map.empty, misnamedConfigs: [], closestPackage: Nothing }
+
+  { loadedPackages } <-
+    State.execStateT (loadSubprojectConfigs rootPath)
+      { loadedPackages: upTree.loadedPackages, blockedSubtrees: [] }
+
+  migrateConfigsWhereNeeded rootPath loadedPackages
```

Comment:
There was a comment about this in the previous version - would be still good to mention that we only migrate the configs after we chose which subtrees/configs to load

Reply:
Added the comment.